laravel/app/EvaluationParser.php

```diff
@@ -14,248 +8,441 @@
 use App\Constant;
 use App\Models\EvaluateData;
 use App\Models\Resident;
-use App\Models\Option;
-use App\Models\Assignment;
 use App\Models\Attending;
 use App\Http\Controllers\MedhubController;
+use Illuminate\Support\Facades\Mail;
+use GuzzleHttp\Client;
+use Carbon\Carbon;
 
-class user{
-    //array 0:LastName, 1:MiddleName, 2:FirstName
-    public $namefl;
-
-    //resident:0, attending:1
-    public $occupation;
-    public $startTime;
-    public $endTime;
-    public $diff;
-    public function __construct($namefl,$occupation,$startTime,$endTime,$diff){
-
-        $this->namefl=$namefl;
-        $this->occupation=$occupation;
-        $this->startTime=$startTime;
-        $this->endTime=$endTime;
-        $this->diff=$diff;
-    }
-
-}
-
-class EvaluationParser extends Model
-{
+class EvaluationParser extends Model {
     protected $filepath;
     protected $date;
-    protected $fileExists = true;
-
-    private static function getDate($line)
-    {
-
-        $month = intval(substr($line, 0, 2));
-        $day = intval(substr($line, 3, 2));
-        $year = intval(substr($line, 6));
-
-        return ($year."-".$month."-".$day);
+    public $residentsFailedAdding = array(); // residents failed to add to DB
+    public $attendingsFailedAdding = array(); // attendings failed to add to DB
+    public $residentsInFile = array(); // all residents that appear in csv file
+    public $attendingsInFile = array(); // all attendings that appear in csv file
+
+    public function __construct($datefile) {
+        $this->filepath = Constant::EVAL_REPORT_PATH . $datefile . Constant::EXTENSION;
+        // get date in yyyy-mm-dd format
+        $year = intval(substr($datefile, 0, 4));
+        $month = intval(substr($datefile, 4, 2));
+        $day = intval(substr($datefile, 6));
+        $this->date = $year . "-" . $month . "-" . $day;
+        self::insertEvaluateData();
```

Comment:
imo, a constructor shouldn't have side effects (inserting the data).

This is especially important with testing, since merely creating an instance of the class does a whole bunch of stuff.

I'd prefer we explicitly call `insertEvaluateData` after creating an instance.

Also, (unimportant, since I think this line should be deleted) shouldn't it be `self->insertEvaluateData();` b/c `insertEvaluateData` isn't a static method?

Reply:
I have changed this so after creating an evaluation parser, `insertEvaluateData` and `notifyForAllFailedUsers` will be called explicitly.